File: R/pullback_vdp.R
# Copyright (c) 2012 Michel Crucifix <michel.crucifix@uclouvain.be>

# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject

# the following conditions:

# The above copyright notice and this permission notice shall be
# incluuded in all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND INFRINGEMENT
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR

# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

# ------------------------------------------------------------------
# R version 2.15.2 (2012-10-26) -- "Trick or Treat"
# ------------------------------------------------------------------ 

require(iceages)

# pullback attractors of vanderpol oscillator
# driven by periodic ersatz of precession and obliquity
# with standard parameters 

Astro <- read_astro(1,1)
times=seq(0,200,0.5)

# determinstic parameter set
parvdp  = c(alpha=30.0, beta=0.7,  gammapre=0.6, gammaobl=0.6,
            omega=4.10, asym=0)

# stochastic parameter set
parvdps = c(alpha=30.0, beta=0.7,  gammapre=0.6, gammaobl=0.6, 
            omega=4.10, sigma=0.5)

# initial conditions positioned on the pullback attractor at time 0
# we do this rather than using the convenient 'pullaback_d' because
# we want to be able to restart from the same IC with the stochastic model

init <- basin(models$vdp_d, par=parvdp, -500., 0, Astro=Astro)$clusters

p41 <- list()
for (i in seq(nrow(init)))
 {
  p41[[i]] <- propagate_d(models$vdp_d, times, init=init[i,], par=parvdp, Astro=Astro)
 }

# tweaks the paramer 
parvdp40 = parvdp
parvdp40['omega'] = 4.0


# generates the 'stochastic' and 'deterministic' tweaked attractors
s41 <- propagate_s(models$vdp_s, init=init[1,], par=parvdps, 
                   times, Astro=Astro, seed=95) 

# we selected here the 'fourth' pullback attractor
# chosen after some trial and error to be representative
# of the phenomenon we want to illustrate

p40 <- pullback_d(models$vdp_d, times=times, par=parvdp40, Astro=Astro)$S[[4]]

# ... and save ! 
save(file='../RData/pullback_11.RData', p41, s41, p40,times)

File: Py/pullback_vdp.py
# Copyright (c) 2012 Michel Crucifix <michel.crucifix@uclouvain.be>

# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject

# the following conditions:

# The above copyright notice and this permission notice shall be
# incluuded in all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND INFRINGEMENT
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR

# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

# ------------------------------------------------------------------
# Code developed for Python 2.7 
# ------------------------------------------------------------------ 

import numpy as np
import matplotlib.pyplot as plt
from matplotlib import mpl
from matplotlib import colors
import itertools

from rpy2 import robjects
file='pullback_34'


fig,ax = plt.subplots(nrows=3,ncols=1)
fig.subplots_adjust( hspace=0.6) 
fig.set_size_inches(7,7)

print(robjects.r.load ('../RData/'+file+'.RData'))
p41 = robjects.r('p41')
s41 = robjects.r('s41')
p40 = robjects.r('p40')
times = np.array(robjects.r('times'))*10.

clr = itertools.cycle(('red','green'))

for Y in p41:
  Y = np.array(Y)[:,0]
  c = clr.next();
  ax[0].plot(times,Y, linewidth=1.2, color=c)
  ax[1].plot(times,Y,'-', linewidth=0.5, color=c)
  ax[2].plot(times,Y,'-', linewidth=0.5, color=c)

ax[2].plot(times,np.array(s41)[:,0], linewidth=1.2, color='black')
ax[1].plot(times,np.array(p40)[:,0], linewidth=1.2, color='black')

ax[0].set_title('pullback attractors $\\tau=41\,\\mathrm{ka}$')
ax[2].set_title('one stochastic realisation  $\\tau=41\,\\mathrm{ka}$')
ax[1].set_title('one pullback attractor $\\tau=40\,\\mathrm{ka}$')

for j in range(3):
  ax[j].xaxis.set_major_locator(plt.MaxNLocator(4))
  ax[j].xaxis.set_major_locator(plt.MaxNLocator(4))
  ax[j].set_xticks(np.arange(0,2000,200))
  ax[j].set_ylabel('y')

ax[2].set_xlabel('time [kyr]')

fig.savefig('../Pdf/pullback_vdp_34.pdf')



